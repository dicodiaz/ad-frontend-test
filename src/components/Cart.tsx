'use client';

import { useCartContext } from '@/context';
import { CartItem } from '.';

export const Cart: React.FC = () => {
  const { cart } = useCartContext();

  return (
    <section className="flex flex-col gap-y-8 px-6 py-8">
      <div className="flex flex-col gap-y-3 text-[#3B3B3B]">
        <h1 className="text-2xl font-bold leading-7 tracking-wide">Your cart</h1>
        <span className="text-xl leading-6 tracking-wide">
          {cart.length} {cart.length === 1 ? 'item' : 'items'}
        </span>
      </div>
      <div className="flex flex-col">
        {cart.map((game) => (
          <CartItem key={game.id} game={game} />
        ))}
      </div>
    </section>
  );
};
