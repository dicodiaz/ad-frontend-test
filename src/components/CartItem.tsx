import { useCartContext } from '@/context';
import { Game } from '@/utils/endpoint';
import { HiX } from 'react-icons/hi';

export type CartItemProps = {
  game: Game;
};

export const CartItem: React.FC<CartItemProps> = ({ game }) => {
  const { id, image, genre, name, price, description, isNew } = game;
  const { dispatch } = useCartContext();

  const removeFromCart = () => {
    dispatch({ type: 'REMOVE_FROM_CART', payload: id });
  };

  return (
    <div className="flex flex-col gap-y-4 px-4 py-5 [&~&]:border-t-[0.5px] [&~&]:border-[#8F8F8F]">
      <div className="relative flex gap-x-3">
        {isNew && (
          <div className="absolute left-3 top-3 rounded bg-[#F5F5F4] px-3 py-2 leading-4 tracking-wide">
            New
          </div>
        )}
        <img src={image} alt={name} className="h-[135px] flex-1 object-cover" />
        <div>
          <button type="button" className="p-1" onClick={removeFromCart}>
            <HiX className="size-5 text-[#8F8F8F]" />
          </button>
        </div>
      </div>
      <div className="flex flex-col gap-y-11">
        <div className="flex flex-col gap-y-3">
          <span className="text-sm font-bold leading-4 text-[#737373]">{genre}</span>
          <div className="flex flex-col gap-y-2">
            <h4 className="text-lg font-bold leading-5 tracking-wide text-[#3B3B3B]">{name}</h4>
            <span className="leading-5 text-[#737373]">{description}</span>
          </div>
        </div>
        <div className="flex justify-end text-lg font-bold leading-5 tracking-wide text-[#3B3B3B]">
          {price.toLocaleString('en-US', { style: 'currency', currency: 'USD' })}
        </div>
      </div>
    </div>
  );
};
