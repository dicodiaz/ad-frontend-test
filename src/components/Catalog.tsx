import { getDataFromServer } from '@/services';
import { GameCard, GenreSelect, SeeMoreButton } from '.';

export type CatalogProps = {
  genre?: string;
  currentPage: number;
};

export const Catalog: React.FC<CatalogProps> = async ({ genre, currentPage }) => {
  const { games, availableFilters, totalPages } = await getDataFromServer({
    genre,
    page: currentPage,
  });

  return (
    <section className="flex flex-col gap-y-2">
      <div className="flex flex-col gap-y-8">
        <GenreSelect availableFilters={availableFilters} />
        <div className="h-px bg-[#EFEDF3]" />
        <div className="flex flex-col gap-y-6 px-6">
          {games.map((game) => (
            <GameCard key={game.id} game={game} />
          ))}
        </div>
      </div>
      {currentPage < totalPages && <SeeMoreButton currentPage={currentPage} />}
    </section>
  );
};
