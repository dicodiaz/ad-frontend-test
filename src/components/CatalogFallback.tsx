import { ImSpinner2 } from 'react-icons/im';

export const CatalogFallback: React.FC = () => {
  return (
    <div className="flex justify-center px-6">
      <ImSpinner2 className="size-6 animate-spin" />
    </div>
  );
};
