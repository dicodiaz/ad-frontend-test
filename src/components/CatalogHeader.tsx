'use client';

import { Data } from '@/services';
import { useRouter, useSearchParams } from 'next/navigation';

export type CatalogHeaderProps = {
  availableFilters: Data['availableFilters'];
};

export const CatalogHeader: React.FC<CatalogHeaderProps> = ({ availableFilters }) => {
  const router = useRouter();
  const searchParams = useSearchParams();

  const handleGenreChange: React.ChangeEventHandler<HTMLSelectElement> = (e) => {
    const genre = e.target.value;
    const params = new URLSearchParams(searchParams);
    if (genre === 'all') {
      params.delete('genre');
    } else {
      params.set('genre', genre);
    }
    router.push(`?${params.toString()}`);
  };

  return (
    <header className="flex flex-col gap-y-8 border-b border-[#EFEDF3] px-6 py-8">
      <h1 className="text-2xl font-bold leading-7 tracking-[0.4px] text-[#3B3B3B]">TOP SELLERS</h1>
      <label className="flex gap-x-6 py-4">
        <span>Genre</span>
        <div className="my-[1px] w-px bg-[#3B3B3B]" />
        <select className="flex-1" onChange={handleGenreChange}>
          <option value="all">All</option>
          {availableFilters
            .sort((a, b) => a.localeCompare(b))
            .map((genre, index) => (
              <option key={index} value={genre}>
                {genre}
              </option>
            ))}
        </select>
      </label>
    </header>
  );
};
