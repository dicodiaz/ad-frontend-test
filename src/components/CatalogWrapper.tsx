import { getDataFromServer } from '@/services';
import { Catalog } from '.';

export type CatalogWrapperProps = {
  genre?: string;
};

export const CatalogWrapper: React.FC<CatalogWrapperProps> = async ({ genre }) => {
  const data = await getDataFromServer({ genre });

  return <Catalog data={data} />;
};
