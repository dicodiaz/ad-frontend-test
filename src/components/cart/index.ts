export * from './Cart';
export * from './CartItem';
export * from './OrderSummary';
