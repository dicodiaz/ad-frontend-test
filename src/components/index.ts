export * from './cart';
export * from './checkout';
export * from './home';
export * from './layout';
