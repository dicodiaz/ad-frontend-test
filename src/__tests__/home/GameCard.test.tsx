import { GameCard } from '@/components';
import { useCartContext } from '@/context';
import { allGames } from '@/utils/endpoint';
import { fireEvent, render, screen } from '@testing-library/react';

jest.mock('@/context/CartContext', () => ({
  useCartContext: jest.fn(),
}));

const mockDispatch = jest.fn();

const mockGame = allGames[0];

const mockCartContextReturnValue = { cart: [], dispatch: mockDispatch, loading: false };

describe('GameCard', () => {
  beforeEach(() => {
    (useCartContext as jest.Mock).mockReturnValue(mockCartContextReturnValue);
  });

  it('renders GameCard unchanged', () => {
    const { container } = render(<GameCard game={mockGame} />);
    expect(container).toMatchSnapshot();
  });

  it('displays the "New" badge if isNew is true', () => {
    const { container } = render(<GameCard game={mockGame} />);
    expect(screen.getByText('New')).toBeInTheDocument();
  });

  it('does not display the "New" badge if isNew is false', () => {
    render(<GameCard game={{ ...mockGame, isNew: false }} />);
    expect(screen.queryByText('New')).not.toBeInTheDocument();
  });

  it('shows "REMOVE" if game is in cart', async () => {
    (useCartContext as jest.Mock).mockReturnValue({
      ...mockCartContextReturnValue,
      cart: [allGames[0]],
    });

    render(<GameCard game={mockGame} />);
    expect(screen.getByRole('button', { name: 'REMOVE' })).toBeInTheDocument();
  });

  it('dispatches ADD_TO_CART when clicked and not in cart', () => {
    render(<GameCard game={mockGame} />);
    fireEvent.click(screen.getByRole('button', { name: 'ADD TO CART' }));
    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'ADD_TO_CART',
      payload: mockGame,
    });
  });

  it('dispatches REMOVE_FROM_CART when clicked and in cart', () => {
    (useCartContext as jest.Mock).mockReturnValue({
      ...mockCartContextReturnValue,
      cart: [allGames[0]],
    });

    render(<GameCard game={mockGame} />);
    fireEvent.click(screen.getByRole('button', { name: 'REMOVE' }));
    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'REMOVE_FROM_CART',
      payload: '1',
    });
  });

  it('disables button and shows spinner when loading', () => {
    (useCartContext as jest.Mock).mockReturnValue({
      ...mockCartContextReturnValue,
      loading: true,
    });

    render(<GameCard game={mockGame} />);
    const button = screen.getByRole('button', { name: 'ADD TO CART' });

    expect(button).toBeDisabled();
    expect(button.querySelector('svg')).toBeInTheDocument();
  });
});
