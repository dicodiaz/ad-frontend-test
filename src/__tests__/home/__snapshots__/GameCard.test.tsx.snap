// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`GameCard renders GameCard unchanged 1`] = `
<div>
  <div
    class="md:w-1/2 md:px-5 xl:w-1/3 xl:flex-none xl:px-6"
    data-testid="game-card"
  >
    <div
      class="border-stroke-secondary flex flex-col gap-y-5 rounded-2xl border-[0.5px] p-6"
    >
      <div
        class="relative overflow-hidden rounded-t-2xl md:h-60"
      >
        <div
          class="text-primary absolute left-3 top-3 rounded bg-stone-100 px-3 py-2 leading-4 tracking-wide"
        >
          New
        </div>
        <img
          alt="Cyberpunk 2077"
          class="h-full object-cover"
          data-nimg="1"
          decoding="async"
          height="240"
          loading="lazy"
          src="/_next/image?url=%2Fgame-images%2Fcyberpunk2077.jpeg&w=1920&q=75"
          srcset="/_next/image?url=%2Fgame-images%2Fcyberpunk2077.jpeg&w=750&q=75 1x, /_next/image?url=%2Fgame-images%2Fcyberpunk2077.jpeg&w=1920&q=75 2x"
          style="color: transparent;"
          width="654"
        />
      </div>
      <div
        class="flex flex-col gap-y-3"
      >
        <span
          class="font-bold text-neutral-500"
        >
          ACTION
        </span>
        <div
          class="flex items-center justify-between gap-x-3 md:min-h-[60px]"
        >
          <h4
            class="text-primary text-lg font-bold leading-5 tracking-wide"
          >
            Cyberpunk 2077
          </h4>
          <span
            class="text-primary text-xl font-bold leading-6 tracking-wide"
          >
            $59.99
          </span>
        </div>
      </div>
      <button
        aria-label="ADD TO CART"
        class="cta-stroke-primary"
        type="button"
      >
        ADD TO CART
      </button>
    </div>
  </div>
</div>
`;
